### ISSUE TITLE
mv empty subfolder causes deadlock

### ISSUE THREAD

**Issue by ktleary** [Microsoft/BashOnWindows#671]
Please use the following bug reporting template to help produce actionable and reproducible issues:
- A brief description
  Bash crashes when trying to move an empty subdirectory - must restart computer (AFAIK) to recover
  e.g.
  root@MACHINE:/mnt/c/bash4/p3/process_content# mv voyage/vc/ .

with vc being an empty subdirectory sends bash into an unrecoverable state
- Expected results
  I would expect it would move the folder or give an error, or respond to cntrl-c
- Actual results (with terminal output if applicable)
  I have to actually restart Windows to  get bash back
- Your Windows build number
  Build 14393.0
- Steps / commands required to reproduce the error
  root@MACHINE:/mnt/c/bash4/p3/process_content# mv voyage/vc/ .
- Strace of the failing command
  N/A
- Required packages and commands to install
  windows ubuntu bash

**Comment by xilun** [comment]
Confirming, also tested on 14393.0

**Comment by benhillis** [comment]
Thanks for reporting this, looks like we have a deadlock due to so incorrect behavior on how we're treating paths with trailing slashes.

This is related to #650

**Comment by xilun** [comment]
Arg deadlocks are not fun. Good luck! ;)

**Comment by lscgzwd** [comment]
When I use mv command in bash. deadlock for me.

**Comment by JasonLinMS** [comment]
Thanks for reporting this issue, this issue has been fixed internally and will be in a future flight.

**Comment by onomatopellan** [comment]
I can confirm it's fixed on build 14901. Thanks!

**Comment by benhillis** [comment]
Fixed in 14901 and later.

**Comment by russalex** [comment]
This one has been marked as fixed for a while. Closing.

**Comment by EvanCarroll** [comment]
I was able to get it to hang just now with a new version of bash trying to rename a directory on the desktop `mv employees_db-full-1.0.6.tar/ employeesdb` You can downplay the file here https://launchpadlibrarian.net/24493586/employees_db-full-1.0.6.tar.bz2

**Comment by benhillis** [comment]
@EvanCarroll - What Windows version are you running?

**Comment by EvanCarroll** [comment]
Windows 10.

**Comment by JasonLinMS** [comment]
Which specific build of Windows 10? If you run winver from the Start menu, it should tell you the specific build on the second line (Version XXXX OS Build XXXXX.XXX).

**Comment by tmcdonnell87** [comment]
Having the same issue: 1607/14393.693

**Comment by StochasticTinkr** [comment]
I'm having the same issue, winver reports 1607/14393.693.

**Comment by JasonLinMS** [comment]
This issue is fixed in 14901 and later, and you guys are on 14393 (Anniversary Update). You would need to be on a Windows Insider build for this to be fixed, or to wait for the Creators Update coming later this year.

**Comment by lkgarrison** [comment]
Thanks for the update @JasonLinMS. Too bad the fix won't come out in a regular windows update :(

**Comment by benhillis** [comment]
@lkgarrison - Fear not, Creator's update is coming soon.

**Comment by whorfin** [comment]
"Don't worry it's fixed in a major update coming soon" is unsatisfactory for a bug resulting in driver/kernel deadlock on user input.
Why is this not an urgent security update for those running lxss?

**Comment by aseering** [comment]
@whorfin -- why would a hang in a default-off and beta-only feature, where the hang only affects that feature, justify an urgent security update?

**Comment by mateusmedeiros** [comment]
@whorfin If you want the last fixes, you have to wait for Creators update or be on an insider build. From what I could gather, WSL is not something modular enough to be updated separately, it's an internal component and is deeply integrated with changes on the Windows kernel.

On Windows insider builds, this has been fixed a long time ago.

**Comment by EvanCarroll** [comment]
If a bug allowed someone to exploit Linux on Windows and cause dead lock would a security update be in order? Or is it just when people run into these issues because of the totally normal lack of interest on Microsoft to actually get this right?

**Comment by aseering** [comment]
@EvanCarroll -- what threat model are you assuming?

**Comment by EvanCarroll** [comment]
I'm not assuming any threat model. I'm posing a hypothetical question. This isn't a mere bug. This was causing a **hard lock** in Windows under regular user-permissions. In such case when the bug is one that can be replicated and surpasses the permissions of the user -- such as forcing a hard-restart or other unrecoverable condition does it really do you any favors to say it doesn't "justify an urgent security update."

The reason why I'm not posing a threat model is because it's blatantly obvious, and I don't want to be insulting. But we can do that. Run an FTP server as a non-admin using Bash on Windows then,

 * Issue an MKDIR, put nothing in the directory.
 * Issue an `RNFR/RNTO` using FTP.

There is a threat model. A remotely exploitable repeatable hard-crash using any FTP server running on Bash on Windows.

**Comment by aseering** [comment]
Thanks for stating the threat model.  A wide variety of people post in this forum; to some, this stuff is obvious, but others don't even understand the concept of a "threat model" to begin with.  Obviously the type of response that's useful for the former case is rather different from what's useful in the latter.

I would clarify a bit:  General *Windows* hard-crashes have, to my knowledge, been getting fixed promptly, even when triggered by WSL.  The concern here is crashes and hangs whose impact is contained within the WSL environment.

In this particular case, I might re-phrase things as "if someone installs and runs an explicitly-unsupported and known-buggy configuration (WSL is [not trying to support daemons yet](https://msdn.microsoft.com/en-us/commandline/wsl/faq#what-is-your-high-priority-feature-set-for-the-windows-10-anniversary-edition), and if you've made a serious effort at running a network service under WSL with any sort of uptime guarantees then you have surely encountered issues) in an environment that is [explicitly labeled as "Beta"](https://msdn.microsoft.com/en-us/commandline/wsl/install_guide), then someone can DoS that service and that entire environment."

I think that's absolutely a valid threat model.  Personally, though, I don't care :-)

I mean, it would be nice, but there are other issues in WSL that I care more about.  Like, for example, implementing the rest of the Linux permissions/security model:  All files accessible from Windows ("DrvFs") are effectively `chmod 777`'ed right now.  Consider someone who serves their Web app out of DrvFs (which has been recommended here on various occasions for interop purposes).  Threat model, attacker who can upload files to your Web service; exploit risk is that they can edit the Web service's code, upload and run Linux executables, even upload and run Windows executables and break out of the WSL sandbox.  Also, as far as I know, many of the ulimit and kernel-namespacing APIs are either unimplemented or stubs; code that depends on those features for security is not secure under WSL.  And so on.

The closest analogy that I can think to draw on the Linux side, though I realize it's imperfect, is to [Wine](https://www.winehq.org/).  You can, if you want to, run Windows system services on Linux under Wine.  But people [don't even bother to report](https://www.cvedetails.com/vulnerability-list/vendor_id-2890/Wine.html) security issues with the Wine environment.  That's not what Wine is about.  And, certainly as of the initial Anniversary-edition release of WSL (which is the only version that contained this particular issue), that wasn't what WSL was about, either.  As WSL continues to mature, perhaps as it gets to the point where the "Beta" label is no longer really needed, I do hope that Microsoft starts patching this type of issue immediately.

**Comment by EvanCarroll** [comment]
Thanks, that's all I was looking for. Bash on Windows can take your system to a grinding halt with a mere FS-command and the level of importance placed on remedying that is roughly summed up with "wait six months and see." That's fine everyone has to start somewhere and most every project at a large corporation starts off as being-unimportant.

Hell, I'm glad Microsoft is doing something but *kernel deadlock on user input* being acceptable doesn't quite raise to the level of maturity that I want.

I understand and reject your analogy to WINE. I understand this is WINE in reverse, but reject that the frame of security-problems (which this would be) introduced by WINE would be treated the same way to Linux distributions. The link provided has [two WINE security vulnerabilities](https://www.cvedetails.com/cve/CVE-2006-0106/).

* CVE-2006-0106, Debian addressed this, patched it, and released the patch the very same month as a security update. See the links to the security update at the [bottom of the page here](https://www.debian.org/security/2006/dsa-954) In addition that design flaw was inherited: *WPM inherits a design flaw from the Windows GDI API, which may lead to the execution of code through GDI escape functions in WMF files.* You can find the bug they [emulated here](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2005-4560).
*  CVE-2005-0787, This was actually patched and released before the CVE was published. You can find the [WINE issue here.](https://bugs.winehq.org/show_bug.cgi?id=2715). It also went out as a [security release.](https://security-tracker.debian.org/tracker/CVE-2005-0787).

In short, things you can do with WINE that you shouldn't be able to do, like kernel dead-lock the system, go out on a fix somewhere, whether it's in an underlying libc bug, a kernel bug, a driver bug, or otherwise.

And, keep in mind WINE -- with one exception -- has only certified one non-game in the Platinum and Gold list. It's predominately used for playing games. That's why I'm excited about this project. Bash on Windows has the potential to bring real software to great gaming platform, albeit one that likes to kernel deadlock.

**Comment by benhillis** [comment]
Some good discussion here, I just want to add a couple of minor points.  Servicing a product like Windows is an extremely difficult and expensive problem.  Every update that is pushed out has to be vetted and analyzed for any potential risk to the install base.  When WSL was initially released we had very few users and was marked as a beta feature.  Both of these factors contribute to the servicing team being less likely to accept the bugs that we wanted to backport.  I acknowledge that we could have probably done a better job of arguing the merits of this particular bug.

Moving forward we re-evaluating what kinds of issues we will backport to the stable version of Windows.  I agree with you, it's unacceptable for a bug like this to go unfixed for 6 months and I will fight to do better in the future.

Here's a rough idea of the types of bugs I think we should be backporting:
1. Bugchecks
2. Security issues
3. Deadlocks or other hangs that cause the driver or usermode service to become unresponsive
4. Targeted syscall bugfixes that unblock large scenarios - this one is very case-by-case.
